File: src/main/java/lantern/Pt.java
package lantern;

import org.eclipse.collections.api.list.ImmutableList;
import org.eclipse.collections.api.factory.Lists;

record Pt(int x, int y) {
    ImmutableList<Pt> adjacent(int width, int height) {
        return Lists.immutable.of(
                new Pt(x - 1, y - 1),
                new Pt(x, y - 1),
                new Pt(x + 1, y - 1),
                new Pt(x + 1, y),
                new Pt(x + 1, y + 1),
                new Pt(x, y + 1),
                new Pt(x - 1, y + 1),
                new Pt(x - 1, y)) //
                .select(p -> p.x >= 0 && p.x < width && p.y >= 0 && p.y < height);
    }
}

File: src/main/kotlin/lantern/Pt.kt
package lantern

data class Pt(val x: Int, val y: Int) {
    fun adjacent(width: Int, height: Int): List<Pt> {
        val xx = listOf(
            Pt(x - 1, y - 1),
            Pt(x, y - 1),
            Pt(x + 1, y - 1),
            Pt(x + 1, y),
            Pt(x + 1, y + 1),
            Pt(x, y + 1),
            Pt(x - 1, y + 1),
            Pt(x - 1, y),
        )
        return xx.filter { p -> (p.x >= 0) && (p.x < width) && (p.y >= 0) && (p.y < height) }
    }
}
